yarn-project/p2p/src/services/tx_collection/tx_collection.ts

```diff
@@ -0,0 +1,253 @@
+  ): Promise<{ txs: TxWithHash[]; requested: TxHash[]; duration: number }>;
+
+  getFastCollectionRequests(): Set<FastCollectionRequest>;
+}
+
+export type CollectionMethod = 'fast-req-resp' | 'fast-node-rpc' | 'slow-req-resp' | 'slow-node-rpc';
+
+export type MissingTxInfo = { blockNumber: number; deadline: Date; readyForReqResp: boolean };
+
+export type FastCollectionRequestInput =
+  | { type: 'block'; block: L2Block }
+  | { type: 'proposal'; blockProposal: BlockProposal };
+
+export type FastCollectionRequest = FastCollectionRequestInput & {
+  missingTxHashes: Set<string>;
+  deadline: Date;
+  blockInfo: BlockInfo;
+  promise: PromiseWithResolvers<void>;
+  foundTxs: Map<string, TxWithHash>;
+};
+
+/**
+ * Coordinates tx collection from remote RPC nodes and reqresp.
+ * The slow collection loops are used for periodically collecting missing txs from mined blocks,
+ * from remote RPC nodes and via reqresp. The fast collection methods are used to quickly gather
+ * txs, usually for attesting to block proposals or preparing to prove an epoch.
+ */
+export class TxCollection {
+  /** Slow collection background loops */
+  private readonly slowCollection: SlowTxCollection;
+
+  /** Fast collection methods */
+  private readonly fastCollection: FastTxCollection;
+
+  /** Loop for periodically reconciling found transactions from the tx pool in case we missed some */
+  private readonly reconcileFoundTxsLoop: RunningPromise;
+
+  /** Metrics */
+  private readonly instrumentation: TxCollectionInstrumentation;
+
+  /** Handler for the txs-added event from the tx pool */
+  private handleTxsAddedToPool: TxPoolEvents['txs-added'] | undefined;
+
+  constructor(
+    private readonly reqResp: Pick<ReqRespInterface, 'sendBatchRequest'>,
+    private readonly nodes: TxSource[],
+    private readonly constants: L1RollupConstants,
+    private readonly txPool: TxPool,
+    private readonly config: TxCollectionConfig,
+    private readonly dateProvider: DateProvider = new DateProvider(),
+    telemetryClient: TelemetryClient = getTelemetryClient(),
+    private readonly log: Logger = createLogger('p2p:tx_collection_service'),
+  ) {
+    const collectionManager: TxCollectionManager = {
+      collect: (collectFn, requested, info) => this.collect(collectFn, requested, info),
+      getFastCollectionRequests: () => this.fastCollection.getFastCollectionRequests(),
+    };
+
+    this.instrumentation = new TxCollectionInstrumentation(telemetryClient, 'TxCollection');
+
+    this.slowCollection = new SlowTxCollection(
+      this.reqResp,
+      this.nodes,
+      collectionManager,
+      constants,
+      this.config,
+      this.dateProvider,
+      this.log,
+    );
+
+    this.fastCollection = new FastTxCollection(
+      this.reqResp,
+      this.nodes,
+      collectionManager,
+      this.config,
+      this.dateProvider,
+      this.log,
+    );
+
+    this.reconcileFoundTxsLoop = new RunningPromise(
```

Comment:
Not sure why we have this, if we listen to the event of adding tx to the pool?

Reply:
Because I have trust issues :-P

If we're lucky, that loop should never do anything.